lib/src/clojure_lsp/feature/diagnostics.clj

```diff
@@ -144,57 +142,73 @@
         (when (not= :unknown (shared/uri->file-type uri))
           (sync-lint-file! uri db))))))
 
-(defn ^:private pmap-light
-  "Call claypoole pmap with less threads than pmap to avoid topping cpu."
-  [f coll]
-  (let [threadpool-size (int (Math/ceil (/ (.. Runtime getRuntime availableProcessors) 3)))]
-    (cp/upmap threadpool-size f coll)))
-
 (defn ^:private unused-public-vars-lint!
-  [definitions project-analysis {:keys [config reg-finding!]} max-parallelize? db]
-  (let [parallelize-fn (if max-parallelize? pmap pmap-light)]
-    (->> definitions
-         (remove (partial exclude-public-diagnostic-definition? config))
-         (parallelize-fn #(when (= 0 (count (q/find-references project-analysis % false db)))
-                            %))
-         (remove nil?)
-         (mapv #(reg-unused-public-var-element! % reg-finding! config))
-         (group-by :filename))))
-
-(defn ^:private project-definitions [project-analysis]
-  (concat (q/find-all-var-definitions project-analysis)
-          (q/find-all-keyword-definitions project-analysis)))
-
-(defn ^:private file-definitions [project-analysis filename]
-  (concat (q/find-var-definitions project-analysis filename false)
-          (q/find-keyword-definitions project-analysis filename)))
+  [var-defs kw-defs project-analysis {:keys [config reg-finding!]}]
+  (let [var-definitions (remove (partial exclude-public-diagnostic-definition? config) var-defs)
+        var-nses (set (map :ns var-definitions)) ;; optimization to limit usages to internal namespaces, or in the case of a single file, to its namespaces
+        var-usage-signature (juxt :to :name)
+        var-usages (into #{}
+                         (comp
+                           (q/xf-all-var-usages-to-namespaces var-nses)
```

Comment:
I wonder if just get all var-usages is faster than filter for those nses, usually all namespaces have var-definitions right? so why do filter those `var-nses` which probably we will get all var-usages in the end?

Reply:
> probably we will get all var-usages in the end?

No, we get many fewer var usages. Remember that though the project analysis has been filtered to internal files, those files use both internal vars _and_ external vars, vars defined in clojure.core and in libraries. Usages of external vars are the majority, so, it's valuable to filter them out.

I tried it both ways as I was developing it, and it's about 25% faster to filter to internal vars, for both single files and for the whole project.

I'd like to document these things, but optimizations are hard to describe and the comments often get out of date as other code changes. Let me know if you'd like to see specific things documented.

Reply:
oh, yes, I agree, but couldn't we do `(->> analysis q/filter-project-only :var-usages)`, we would get all var-usages from project-only too right?

Reply:
> couldn't we do `(->> analysis q/filter-project-only :var-usages)`

First, let's make sure we're on the same page. I think this is what you meant:

```clojure
(->> analysis                               ;; 1
     q/filter-project-analysis              ;; 2
     (mapcat val)                           ;; 3
     (filter #(= :var-usages (:bucket %)))) ;; 4
```

Is that right? You wrote `q/filter-project-only`. Is that something I don't know about, or did you mean `q/filter-project-analysis`?

Again, the answer is no, that's not the same. Let's go line-by-line. Suppose we start with this `analysis`, line 1:

```clojure
{"clojure_lsp/shared.clj"  '[{:bucket :var-definitions
                              :ns clojure-lsp.shared
                              :name inside?}
                             {:bucket :var-usages  ;; <<< we don't care about this var-usage
                              :to clojure.core
                              :name inc}]
 "clojure_lsp/queries.clj" '[{:bucket :var-definitions
                              :ns clojure-lsp.queries
                              :name find-local-usages-under-form}
                             {:bucket :var-usages
                              :to clojure-lsp.shared
                              :name inside?}]
 "clojure/core.clj"        '[{:bucket :var-definitions
                              :ns clojure.core
                              :name inc}]}
```
Then `q/filter-project-analysis`, line 2:
```clojure
{"clojure_lsp/shared.clj"  '[{:bucket :var-definitions
                              :ns clojure-lsp.shared
                              :name inside?}
                             {:bucket :var-usages  ;; even though we've filtered to project analysis, we still have this var-usage
                              :to clojure.core
                              :name inc}]
 "clojure_lsp/queries.clj" '[{:bucket :var-definitions
                              :ns clojure-lsp.queries
                              :name find-local-usages-under-form}
                             {:bucket :var-usages
                              :to clojure-lsp.shared
                              :name inside?}]}
```
"clojure/core.clj" has been removed, but we still have **usages** of clojure.core. Then `(mapcat val)`, line 3:
```clojure
'[{:bucket :var-definitions
   :ns clojure-lsp.shared
   :name inside?}
  {:bucket :var-usages
   :to clojure.core
   :name inc}
  {:bucket :var-definitions
   :ns clojure-lsp.queries
   :name find-local-usages-under-form}
  {:bucket :var-usages
   :to clojure-lsp.shared
   :name inside?}]
```
Then `(filter #(= :var-usages (:bucket %)))`, line 4:
```clojure
'[{:bucket :var-usages ;; still here
   :to clojure.core
   :name inc}
  {:bucket :var-usages
   :to clojure-lsp.shared
   :name inside?}]
```
At this point we know that `clojure.core/inc` and `clojure-lsp.shared/inside?` are both used. But it's not important that `clojure.core/inc` is used! My code filters **usages to** internal namespaces:
```clojure
'[{:bucket :var-usages
   :to clojure-lsp.shared
   :name inside?}]
```

This list is smaller, and makes the rest of the code faster.

Reply:
Got it, makes total sense, thanks for the detailed explanation, even having var-usages from project they could be usages of external definitions which we don't care, I can see that now